File: content/courses/parallel-computing-introduction/codes/omp_for.f90

use omp_lib

   integer, parameter    :: M=10;
   integer, dimension(M) :: a, b
   real,    dimension(M) :: c
   integer               :: i,j
   integer               :: low, high

   a=[1,3,5,7,2,3,4,5,6,7]
   b=[9,5,7,9,5,7,8,9,10,4]
   c=0.0

   !$omp parallel 

   do i=1,M
      low = a(i)
      high = b(i)

      if (low > high) then
         write(*,*) "Exiting ",i
         exit
      endif

      !$omp do private(j)
      do j=low,high
         c(i) = c(i) - real(a(j))/real(b(j))
      enddo
      !$omp end do

   enddo
   !$omp end parallel

   do i=1,M
      write(*,'(a,i4,a,g12.6)') "i=",i,"  c(i)=",c(i)
   enddo

end program

File: content/courses/parallel-computing-introduction/codes/omp_for.py
from omp4py import *
import numpy as np

@omp
def parparfor(M):
   a=np.array([0,2,4,6,1,2,3,4,5,6])
   b=np.array([9,4,8,8,4,6,7,8,9,3])
   c=np.zeros(M)

   with omp("parallel"):
       for i in range(M):
           low = a[i]
           high = b[i]

           if low > high:
              print (f"Exiting {i}")
              break

           j=0
           with omp("for private(j)"):
               for j in range(low,high+1):
                   print(i,c[i])
                   c[i] -= a[j]/b[j]

   for i in range(M):
       print (f"i={i} c[i]={c[i]}")

parparfor(10)

